### DISCUSSION TITLE
Scripting language?

### DISCUSSION THREAD

**Discussion by Racso** [o3de/o3de#1888] [Q&A]
Hello! I have two questions regarding o3de:

1. What's the scripting language for o3de? Lua?
2. In case it's Lua: Are there plans to support compiled and/or statically-typed languages such as C# or C++?

Thanks!

**Comment by Racso** [comment]
Update [from the docs](https://o3de.org/docs/user-guide/scripting/lua/) regarding my first question:

> O3DE uses Lua version 5.3 .

**Comment by Dimous** [comment]
Hey! May I chime in?
If there was Lua api reference, someone would create Haxe externs like, for example: [Lua] https://github.com/hxdefold/hxdefold or [C#] https://github.com/HaxeGodot/godot or even [C++] https://github.com/proletariatgames/unreal.hx. 
`Haxe is an open source high-level strictly-typed programming language with a fast optimizing cross-compiler.`
https://haxe.org/manual/target-lua-getting-started.html
It is quite popular in gamedev.

#idea wink-wink

**Reply by Dimous** [reply]
Take another example -- https://github.com/OTRD5k/RBXHaxe 
Roblox Haxe externs are generated automatically based on consumable Lua api reference -- http://setup.roblox.com/version-8e89b15fa1e44a8d-API-Dump.json
If only o3de had something similar.

**Reply by lsemp3d** [reply]
It's possible to generate an API document by traversing through the Behavior Context and writing out its contents, additionally, because the API is shared by Script Canvas it would also generate useful documentation for it. It's a great idea!

**Reply by Infurael** [reply]
The Serialize Context Tools is already capable of exporting the Serialize Context and Edit Context to a JSON file that can in turn be used to generate documentation from. The Behavior Context isn't yet supported, but it's pretty similar to the Serialize Context.

**Reply by troys-code** [reply]
This is interesting, but can I take 10,000 lines of Unity C# code and port it to Haxe/C# without much hassle? @Dimous 

Seems like the language is limited compared to something like dotnet (cross-platform C# standard library).

**Reply by Dimous** [reply]
@lsemp, @AMZN-koppersr thanks for the tip!
@troy-lamerton try this -- https://github.com/FizzerWL/Cs2hx
`CS2HX aims to support 90% of an average C# codebase. You should expect to have to re-factor some of your code to avoid features that CS2HX does not support.`
